### ISSUE TITLE
trick-ICG won't build on Redhat 7.X w/ gcc 5.0+ and clang 3.4.2 without -D_GLIBCXX_USE_CXX11_ABI=0

### ISSUE THREAD

**Issue by ddj116** [nasa/trick#1253]
# How to replicate

This is replicated in the FSL with the following machine config:
```
> which gcc && gcc --version
/software/x86_64/gcc/8.3.0/bin/gcc
gcc (GCC) 8.3.0
> which llvm-config && llvm-config --version
/usr/bin/llvm-config
3.4.2
> which clang && clang --version
/usr/bin/clang
clang version 3.4.2 (tags/RELEASE_34/dot2-final)
Target: x86_64-redhat-linux-gnu
Thread model: posix
> cat /etc/redhat-release 
CentOS Linux release 7.9.2009 (Core)
> rpm -qa |grep llvm
llvm-private-7.0.1-1.el7.x86_64
llvm-3.4.2-9.el7.x86_64
llvm-libs-3.4.2-9.el7.x86_64
llvm-static-3.4.2-9.el7.x86_64
llvm-devel-3.4.2-9.el7.x86_64
```
To replicate the error, run `./configure && make` which will result in many undefined references when `trick-ICG` attempts to link:

```
/software/x86_64/gcc/8.3.0/bin/g++ -o /local/user/GGRoLS2/trick/bin/trick-ICG object_Linux_8.3_x86_64/ClassVisitor.o object_Linux_8.3_x86_64/PrintFileContents10.o object_Linux_8.3_x86_64/ICGASTConsumer.o object_Linux_8.3_x86_64/ClassValues.o object_Linux_8.3_x86_64/Utilities.o object_Linux_8.3_x86_64/TypedefVisitor.o object_Linux_8.3_x86_64/EnumVisitor.o object_Linux_8.3_x86_64/FieldVisitor.o object_Linux_8.3_x86_64/PrintAttributes.o object_Linux_8.3_x86_64/FindTrickICG.o object_Linux_8.3_x86_64/VariableVisitor.o object_Linux_8.3_x86_64/ConstructValues.o object_Linux_8.3_x86_64/EnumValues.o object_Linux_8.3_x86_64/HeaderSearchDirs.o object_Linux_8.3_x86_64/main.o object_Linux_8.3_x86_64/PrintFileContentsBase.o object_Linux_8.3_x86_64/CommentSaver.o object_Linux_8.3_x86_64/TranslationUnitVisitor.o object_Linux_8.3_x86_64/FieldDescription.o object_Linux_8.3_x86_64/ClassTemplateVisitor.o object_Linux_8.3_x86_64/map_trick_units_to_udunits.o -lclangFrontend -lclangDriver -lclangSerialization -lclangParse -lclangSema -lclangAnalysis -lclangEdit -lclangAST -lclangLex -lclangBasic -lLLVMInstrumentation -lLLVMIRReader -lLLVMAsmParser -lLLVMDebugInfo -lLLVMOption -lLLVMLTO -lLLVMLinker -lLLVMipo -lLLVMVectorize -lLLVMBitWriter -lLLVMBitReader -lLLVMTableGen -lLLVMR600CodeGen -lLLVMR600Desc -lLLVMR600Info -lLLVMR600AsmPrinter -lLLVMSystemZDisassembler -lLLVMSystemZCodeGen -lLLVMSystemZAsmParser -lLLVMSystemZDesc -lLLVMSystemZInfo -lLLVMSystemZAsmPrinter -lLLVMHexagonCodeGen -lLLVMHexagonAsmPrinter -lLLVMHexagonDesc -lLLVMHexagonInfo -lLLVMNVPTXCodeGen -lLLVMNVPTXDesc -lLLVMNVPTXInfo -lLLVMNVPTXAsmPrinter -lLLVMCppBackendCodeGen -lLLVMCppBackendInfo -lLLVMMSP430CodeGen -lLLVMMSP430Desc -lLLVMMSP430Info -lLLVMMSP430AsmPrinter -lLLVMXCoreDisassembler -lLLVMXCoreCodeGen -lLLVMXCoreDesc -lLLVMXCoreInfo -lLLVMXCoreAsmPrinter -lLLVMMipsDisassembler -lLLVMMipsCodeGen -lLLVMMipsAsmParser -lLLVMMipsDesc -lLLVMMipsInfo -lLLVMMipsAsmPrinter -lLLVMARMDisassembler -lLLVMARMCodeGen -lLLVMARMAsmParser -lLLVMARMDesc -lLLVMARMInfo -lLLVMARMAsmPrinter -lLLVMAArch64Disassembler -lLLVMAArch64CodeGen -lLLVMAArch64AsmParser -lLLVMAArch64Desc -lLLVMAArch64Info -lLLVMAArch64AsmPrinter -lLLVMAArch64Utils -lLLVMPowerPCCodeGen -lLLVMPowerPCAsmParser -lLLVMPowerPCDesc -lLLVMPowerPCInfo -lLLVMPowerPCAsmPrinter -lLLVMSparcCodeGen -lLLVMSparcDesc -lLLVMSparcInfo -lLLVMX86Disassembler -lLLVMX86AsmParser -lLLVMX86CodeGen -lLLVMSelectionDAG -lLLVMAsmPrinter -lLLVMX86Desc -lLLVMX86Info -lLLVMX86AsmPrinter -lLLVMX86Utils -lLLVMMCDisassembler -lLLVMMCParser -lLLVMInterpreter -lLLVMMCJIT -lLLVMJIT -lLLVMCodeGen -lLLVMObjCARCOpts -lLLVMScalarOpts -lLLVMInstCombine -lLLVMTransformUtils -lLLVMipa -lLLVMAnalysis -lLLVMRuntimeDyld -lLLVMExecutionEngine -lLLVMTarget -lLLVMMC -lLLVMObject -lLLVMCore -lLLVMSupport -L/usr/lib64/llvm  -lz -lpthread -lffi -ltinfo -ldl -lm  -ludunits2
object_Linux_8.3_x86_64/ClassVisitor.o: In function `clang::NamedDecl::getNameAsString[abi:cxx11]() const':
/usr/include/clang/AST/Decl.h:144: undefined reference to `clang::DeclarationName::getAsString[abi:cxx11]() const'
```
If you look at `/usr/lib64/llvm` they indeed exist, and `nm /usr/lib64/llvm/libclangAST.a --demangle |grep clang::DeclarationName::getAsString` shows the symbol is there. 

# My workaround

The clue is the `abi:cxx11` part in the `undefined reference` error. Although the symbols exist in the library, they aren't a compatible abi tag so the linker can't find what it's looking for.  According to link (2) in the documentation below, a change in abi tags came in `gcc` 5.0 and later and older libraries built with `gcc` < 5.0 will not work with newer gcc versions unless `-D_GLIBCXX_USE_CXX11_ABI=0` is given in the `gcc` compilation line.

In order to get the prerequisites of `trick-ICG` to build with these flags, I had to make two changes.
1. Add `-D_GLIBCXX_USE_CXX11_ABI=0` to `CXXFLAGS` in `trick_source/codegen/Interface_Code_Gen/makefile`
2. Manually run the `g++` line for `map_trick_units_to_udunits.o` with `-D_GLIBCXX_USE_CXX11_ABI=0` added. I couldn't find where in the makefiles to add this flag for this udunits object.

After (1) and (2), `cd trick && make` will finish.  @jmpenn @spfennell is the proper solution to add this flag to the Trick build? Is there a better way to implement it?  FYI @alexlin0 @dbankieris for awareness.

# Documentation

1. Similar issues #174, #259
4. Explanation of the abi and backwards compatibility: https://developers.redhat.com/blog/2015/02/05/gcc5-and-the-c11-abi
5. https://bugs.llvm.org/show_bug.cgi?id=23529
6. https://gcc.gnu.org/legacy-ml/gcc-patches/2015-04/msg00153.html

**Comment by ddj116** [comment]
Per discussion today in person, we also believe that this could be resolved by FSL moving to the "developer toolset" versions of `gcc` as [described here](https://stackoverflow.com/a/58650444). I'll be working with the admins on this path, in the meantime the ability to pass in the `-D_GLIBCXX_USE_CXX11_ABI=0` flag to the relevant files without modifying Trick would help us workaround this issue in the short term.

**Comment by ddj116** [comment]
Thanks for the input @iamthad - the magic switch to `gcc`'s `./configure` which gets us past this issue is ` --with-default-libstdcxx-abi=gcc4-compatible` which is now implemented in a version of gcc 8.3.0 in our lab.

Closing this since this isn't really an issue with Trick. Hopefully this thread will help others that may run into this.
